src/server/service/kv.rs

```diff
@@ -1450,31 +1450,79 @@ fn future_raw_delete_range<E: Engine, L: LockManager, P: PdClient + 'static>(
     })
 }
 
-// unimplemented
 fn future_ver_get<E: Engine, L: LockManager, P: PdClient + 'static>(
-    _storage: &Storage<E, L, P>,
-    mut _req: VerGetRequest,
+    storage: &Storage<E, L, P>,
+    mut req: VerGetRequest,
 ) -> impl Future<Item = VerGetResponse, Error = Error> {
-    let resp = VerGetResponse::default();
-    future::ok(resp)
+    let mut ver_value = VerValue::default();
+    let version: TimeStamp = req.get_start_version().into();
+    storage
+        .get(
+            req.take_context(),
+            Key::from_raw(req.get_key()),
+            version.clone(),
+        )
+        .then(move |v| {
+            let mut resp = VerGetResponse::default();
+            if let Some(err) = extract_region_error(&v) {
+                resp.set_region_error(err);
+            } else {
+                match v {
+                    Ok(Some(val)) => {
+                        ver_value.set_value(val);
+                        ver_value.set_version(version.physical());
+                        resp.set_value(ver_value);
+                    }
+                    Ok(None) => resp.set_not_found(true),
+                    Err(e) => resp.set_error(extract_ver_error(&e)),
+                }
+            }
+            Ok(resp)
+        })
 }
 
-// unimplemented
 fn future_ver_batch_get<E: Engine, L: LockManager, P: PdClient + 'static>(
-    _storage: &Storage<E, L, P>,
-    mut _req: VerBatchGetRequest,
+    storage: &Storage<E, L, P>,
+    mut req: VerBatchGetRequest,
 ) -> impl Future<Item = VerBatchGetResponse, Error = Error> {
-    let resp = VerBatchGetResponse::default();
-    future::ok(resp)
+    // TODO: modify this to get several versions of one key
+    let keys = req.get_key().iter().map(|x| Key::from_raw(x)).collect();
+    storage
+        .batch_get(req.take_context(), keys, req.get_start_version().into())
+        .then(|v| {
+            let mut resp = VerBatchGetResponse::default();
+            if let Some(err) = extract_region_error(&v) {
+                resp.set_region_error(err);
+            } else {
+                resp.set_pairs(extract_verkv_pairs(v).into());
+            }
+            Ok(resp)
+        })
 }
 
-// unimplemented
 fn future_ver_mut<E: Engine, L: LockManager, P: PdClient + 'static>(
```

Comment:
`ver_mut` has two semantics `put` and `delete`, and for verkv the data is put into another verkv CF

Reply:
I see